firedrake/slate/slac/compiler.py

```diff
@@ -0,0 +1,324 @@
+linear algebra operations are performed using the `Eigen::Matrix` methods built into Eigen.
+"""
+from __future__ import absolute_import, print_function, division
+
+import petsc
+
+from coffee import base as ast
+
+from firedrake.constant import Constant
+from firedrake.tsfc_interface import SplitKernel, KernelInfo
+from firedrake.slate.slate import (TensorBase, Tensor,
+                                   Transpose, Inverse, Negative,
+                                   Add, Sub, Mul, Action)
+from firedrake.slate.slac.kernel_builder import KernelBuilder
+from firedrake import op2
+
+from tsfc.parameters import SCALAR_TYPE
+
+from ufl.coefficient import Coefficient
+
+
+__all__ = ['compile_expression']
+
+
+def compile_expression(slate_expr, tsfc_parameters=None):
+    """Takes a SLATE expression `slate_expr` and returns the appropriate
+    :class:`firedrake.op2.Kernel` object representing the SLATE expression.
+
+    :arg slate_expr: a :class:'TensorBase' expression.
+
+    :arg tsfc_parameters: an optional `dict` of form compiler parameters to
+                          be passed onto TSFC during the compilation of ufl forms.
+    """
+    if not isinstance(slate_expr, TensorBase):
+        raise ValueError("Expecting a `slate.TensorBase` expression, not a %r" % slate_expr)
+
+    # TODO: Get PyOP2 to write into mixed dats
+    if any(len(a.function_space()) > 1 for a in slate_expr.arguments()):
+        raise NotImplementedError("Compiling mixed slate expressions")
+
+    # Initialize shape and statements list
+    shape = slate_expr.shape
+    statements = []
+
+    # Create a builder for the SLATE expression
+    builder = KernelBuilder(expression=slate_expr, tsfc_parameters=tsfc_parameters)
+
+    # Initialize coordinate and facet symbols
+    coordsym = ast.Symbol("coords")
+    coords = None
+    cellfacetsym = ast.Symbol("cell_facets")
+    inc = []
+
+    # Now we construct the list of statements to provide to the builder
+    context_temps = builder.temps.copy()
+    for exp, t in context_temps.items():
+        statements.append(ast.Decl(eigen_matrixbase_type(exp.shape), t))
+        statements.append(ast.FlatBlock("%s.setZero();\n" % t))
+
+        for splitkernel in builder.kernel_exprs[exp]:
+            clist = []
+            index = splitkernel.indices
+            kinfo = splitkernel.kinfo
+            integral_type = kinfo.integral_type
+
+            if integral_type not in ["cell", "interior_facet", "exterior_facet"]:
+                raise NotImplementedError("Integral type %s not currently supported." % integral_type)
+
+            coordinates = exp.ufl_domain().coordinates
+            if coords is not None:
+                assert coordinates == coords
+            else:
+                coords = coordinates
+
+            for cindex in kinfo.coefficient_map:
+                c = exp.coefficients()[cindex]
+                clist.append(builder.coefficient_map()[c])
```

Comment:
Ah, I don't think I do anything special to address this case. This is simply a map: `dict((c, ast.Symbol("w%d" % i)) for i, c in enumerate(self.expression.coefficients()))`